### PR TITLE
Feature/arcii model

### PR DESCRIPTION
**Pull request by pl8787**

### CODE REVIEW

**Comment by faneshion** [comment]
You can add the mask into the model since we have already provided length information of each text.

**Comment by pl8787** [comment]
> You can add the mask into the model since we have already provided length information of each text.

OK

File: matchzoo/models/arcii_model.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        >>> model.build()
+
+    """
+
+    @classmethod
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params()
+        params['optimizer'] = 'adam'
+        params['input_shapes'] = [(32,), (32,)]
```
Comment (uduse) [inline matchzoo/models/arcii_model.py]: `input_shapes` better to be left empty. It should be filled by either `guess_and_fill_missing_parameters` or manually by users.

File: matchzoo/models/arcii_model.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    @property
+    def embedding_mat(self) -> np.ndarray:
+        """Get pretrained embedding for ArcII model."""
+        # Check if provided embedding matrix
+        if self._params['embedding_mat'] is None:
+            s = self._params['embedding_random_scale']
+            self._params['embedding_mat'] = \
+                np.random.uniform(-s, s, (self._params['vocab_size'],
+                                          self._params['embedding_dim']))
+        return self._params['embedding_mat']
```
Comment (uduse) [inline matchzoo/models/arcii_model.py]: Things in `self._params` should not be altered here, or it will cause problems when saving and loading. Store the embedding matrix in `self._embedding_matrix` or something.

File: matchzoo/layers/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from .match_layer import MatchLayer
+
+layer_dict = {
```
Comment (bwanglzu) [inline matchzoo/layers/__init__.py]: can you rephrase the variable name?

File: matchzoo/layers/match_layer.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
+"""An implementation of Match Layer."""
+from __future__ import absolute_import
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: remove

File: matchzoo/layers/match_layer.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+import typing
+from keras import backend as K
+from keras.engine import Layer
+
+
+class MatchLayer(Layer):
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: also it's a bit ambiguous, since any layer could be a Matching Layer, for example, MultiPerspectiveMatching is more descriptive.

File: matchzoo/layers/match_layer.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    def __init__(self, normalize: bool = False,
+                 match_type: str = 'dot', **kwargs):
+        """
+        :class:`MatchLayer` constructor.
+
+        :param normalize: Whether to L2-normalize samples along the
+            dot product axis before taking the dot product.
+            If set to True, then the output of the dot product
+            is the cosine proximity between the two samples.
+        :param **kwargs: Standard layer keyword arguments.
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: I think you can add the `match_type` and remove `**kwargs`

File: matchzoo/layers/match_layer.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        """
+        super(MatchLayer, self).__init__(**kwargs)
+        self.normalize = normalize
+        self.match_type = match_type
+        self.supports_masking = True
+        if match_type not in ['dot', 'mul', 'plus', 'minus', 'concat']:
+            raise ValueError('In `MatchLayer` layer, '
+                             'match_type={} is unknown.'.format(match_type))
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: ValueError("{} is not a valid matching type, {} expected.")

File: matchzoo/layers/match_layer.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        """
+        # Used purely for shape validation.
+        if not isinstance(input_shape, list) or len(input_shape) != 2:
+            raise ValueError('A `MatchLayer` layer should be called '
+                             'on a list of 2 inputs.')
+        self.shape1 = input_shape[0]
+        self.shape2 = input_shape[1]
+        if self.shape1[0] != self.shape2[0]:
+            raise ValueError(
+                'Dimension incompatibility '
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: dimensional incompatibility or incompatible dimensions, personally think the later is better.

File: matchzoo/layers/match_layer.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+                '{} != {}. Layer shapes: {}, {}.'.format(self.shape1[0],
+                                                         self.shape2[0],
+                                                         self.shape1,
+                                                         self.shape2))
+        if self.shape1[2] != self.shape2[2]:
+            raise ValueError(
+                'Dimension incompatibility '
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: same

File: matchzoo/layers/match_layer.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+    def call(self, inputs: list) -> typing.Any:
+        """
+        The computation logic of MatchLayer.
+
+        :param inputs: two input tensors.
+        """
+        x1 = inputs[0]
+        x2 = inputs[1]
+        if self.match_type in ['dot']:
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: self.match_type is a string, then use `if self._match_type == 'dot'`

File: matchzoo/models/arcii_model.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
+"""An implementation of ArcII Model."""
+from matchzoo import engine
+from matchzoo import layers
+import typing
+import numpy as np
+from keras.models import Model
+from keras.layers import Input, Embedding, Conv1D, Conv2D, MaxPooling2D, \
```
Comment (bwanglzu) [inline matchzoo/models/arcii_model.py]: see #351

File: matchzoo/models/arcii_model.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        # Phrase level representations
+        conv_1d_left = Conv1D(self._params['kernel_1d_count'],
+                              self._params['kernel_1d_size'],
+                              padding=self._params['padding'])(embed_left)
+        conv_1d_right = Conv1D(self._params['kernel_1d_count'],
+                               self._params['kernel_1d_size'],
+                               padding=self._params['padding'])(embed_right)
+
+        # Interaction
+        embed_cross = layers.MatchLayer(match_type='plus')([conv_1d_left,
```
Comment (bwanglzu) [inline matchzoo/models/arcii_model.py]: can you rephrase the variable name?

File: matchzoo/layers/match_layer.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            elif self.match_type == 'plus':
+                output = x1_exp + x2_exp
+            elif self.match_type == 'minus':
+                output = x1_exp - x2_exp
+        elif self.match_type in ['concat']:
+            x1_exp = K.tf.stack([x1] * self.shape2[1], axis=2)
+            x2_exp = K.tf.stack([x2] * self.shape1[1], axis=1)
+            output = K.tf.concat([x1_exp, x2_exp], axis=3)
+
+        return output
```
Comment (bwanglzu) [inline matchzoo/layers/match_layer.py]: line 59-87 is not readable.
